Add Transport for London integration

pwheel opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
Adds a Transport for London (TfL) integration to Home Assistant.
The integration adds a sensor for arrivals for services (bus, tube, rail, etc) to given stops.
It has similarities to:

1. [london_underground](https://github.com/home-assistant/core/tree/dev/homeassistant/components/london_underground) integration - which allows to track line status of london underground lines
2. [london_tfl custom component](https://github.com/morosanmihail/HA-LondonTfL) integration - which provides the same functionality but is a custom component

This integration could be extended to add functionality from london_underground in future.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/30007

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

pwheel committed:
Fairly complete version of the TfL integration. Neem more tests for sensor.py

pwheel committed:
Merge branch 'dev' into tfl-integration

pwheel committed:
Added more tests to the TfL integration for the sensor and options flow.

pwheel committed:
Added tests for connection errors during startup

pwheel committed:
Tfl - tweaked the options flow so the dialog has the right text for stops.

pwheel committed:
Merge branch 'dev' into tfl-integration

pwheel committed:
Merge branch 'dev' into tfl-integration

pwheel committed:
Removed commented out code from tfl integration

pwheel committed:
- Added __init.py__ for tests/components/tfl to ensure CODEOWNERS is generated correctly
- Updated manifest.json
- Reran script.hassfest

pwheel committed:
Merge branch 'dev' into tfl-integration

pwheel committed:
Merge branch 'dev' into tfl-integration

joostlek commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,230 @@
+        vol.Optional(CONF_API_APP_KEY, default=""): cv.string,
+    }
+)
+STEP_STOP_POINT_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_STOP_POINT): cv.string,
+        vol.Optional(CONF_STOP_POINT_ADD_ANOTHER): cv.boolean,
+    }
+)
```
Wouldn't a select selector be good here? They allow custom values

pwheel replied:
Thanks for feedback!
Just had a quick play around with TextSelector supporting multiple, it does make for a nice UX yes.
At the moment the code is validating each stop point (i.e. that it exists) one at a time when the form is submitted which makes the error handling easy if a stop point doesn't exist.
I'll play around with what can be done to give good feedback in that scenario. Dunno if it's possible to give feedback on which stop point was not valid with the TextSelector approach.

pwheel replied:
Resolved this today, thanks

emontnemery commented on homeassistant/components/tfl/__init__.py:
```diff
@@ -0,0 +1,83 @@
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+_STOP_POINT_SCHEME = vol.Schema({vol.Required(CONF_ID): cv.string})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {vol.Required(CONF_STOP_POINTS): [_STOP_POINT_SCHEME]}
+)
+
+
```
Please remove this, it's not needed

pwheel replied:
Done! Thank you

emontnemery commented on homeassistant/components/tfl/config_helper.py:
Please remove this module, it's not needed

pwheel replied:
Done

emontnemery commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,230 @@
+
+                # return await self.async_step_stop_points()
+                return self.async_create_entry(
+                    title="Transport for London", data=self.data
```
Instead of storing all the user input in config entry data, and then copying it to user entry option in the options flow, store the settings which the user should be able to change in options already in the config flow.

pwheel replied:
Ah, definitely felt I was missing something here. Docs and tutorials I'd seen tend to keep the data separate, which led me in the wrong direction.
Your suggestion is… obvious in hindsight. Thanks!

pwheel replied:
Done

emontnemery commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,230 @@
+
+        options_schema = vol.Schema(
+            {
+                vol.Optional(CONF_API_APP_KEY, default=api_key): cv.string,
```
Why do we allow the user to change the app key from an options flow?
If we can detect the app key is invalidated, we should instead trigger a reauth flow

pwheel replied:
This is because I didn't know there was a `reconfigure` flow.

I'll take a look at it

pwheel replied:
@emontnemery I've actioned this so it now uses a reconfigure flow

emontnemery commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,230 @@
+        vol.Optional(CONF_API_APP_KEY, default=""): cv.string,
+    }
+)
+STEP_STOP_POINT_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_STOP_POINT): cv.string,
+        vol.Optional(CONF_STOP_POINT_ADD_ANOTHER): cv.boolean,
+    }
+)
```
I agree with the comment from @joostlek, a select selector allowing the user to add multiple stops may work better.

pwheel replied:
Done

emontnemery reviewed (changes requested):
Some initial comments

pwheel committed:
Changed to use a TextSelector for the Options flow and improved some error handling

pwheel committed:
Removed unneeded code

pwheel committed:
Store data in options during create so can remove config_helper

pwheel committed:
Merge branch 'dev' into tfl-integration

pwheel committed:
Merge branch 'dev' into tfl-integration

pwheel committed:
Moved the app_key out of options and to the reconfigure flow

pwheel commented:
@emontnemery , @joostlek - thanks very much for your feedback; I've actioned all of it, please take a look and let me know your feedback. Thanks!

pwheel committed:
Merge branch 'dev' into tfl-integration

emontnemery commented on homeassistant/components/tfl/__init__.py:
```diff
@@ -0,0 +1,74 @@
+            raise ConfigEntryAuthFailed(
+                "Authentication failure for app_key=" + app_key
+            ) from exception
+
+        raise ConfigEntryNotReady(
+            "Connection error whilst connecting to TfL"
+        ) from exception
+    except URLError as exception:
```
The position of the blank line is a bit strange to my eyes. Maybe do:

```suggestion

        raise ConfigEntryNotReady(
            "Connection error whilst connecting to TfL"
        ) from exception

    except URLError as exception:
```

emontnemery commented on homeassistant/components/tfl/__init__.py:
```diff
@@ -0,0 +1,74 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Transport for London from a config entry."""
+
+    hass.data.setdefault(DOMAIN, {})
```
Instead of this, please use this pattern https://developers.home-assistant.io/blog/2024/04/30/store-runtime-data-inside-config-entry

pwheel replied:
updated

emontnemery commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,284 @@
+        # TfL's API returns a 429 if you pass an invalid app_key, but we also check
+        # for other reasonable error codes in case their behaviour changes
+        error_code = exception.code
+        if error_code in (429, 401, 403):
```
The same error codes are also in `__init__.py`, store the tuple in `const.py` instead and use it in both places

pwheel replied:
Tidied up as part of the wider cleanup you called out with the repeated error handling. Only there once now, so I didn't move out to `const.py`

emontnemery commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,284 @@
+    async def async_step_reconfigure(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Add reconfigure step to allow to reconfigure a config entry."""
```
Consider changing the docstring to explain what *this* integration's reconfigure flow allows

emontnemery commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,284 @@
+        )
+        assert config_entry is not None
+        if user_input is not None:
+            try:
+                await validate_app_key(self.hass, user_input[CONF_API_APP_KEY])
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
```
This pattern is repeated several times: here , in the user step and in the `validate_stop_points` helper. Try to refactor so you don't have to repeat yourself.

emontnemery commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,284 @@
+
+        return self.async_show_form(
+            step_id="reconfigure",
+            data_schema=vol.Schema(
+                {
+                    vol.Optional(
+                        CONF_API_APP_KEY, default=config_entry.data[CONF_API_APP_KEY]
+                    ): cv.string,
+                }
+            ),
```
Instead of this, use a constant schema and call `add_suggested_values_to_schema`:

```py
data_schema=self.add_suggested_values_to_schema(
    RECONFIGURE_SCHEMA, config_entry.data
)
```

emontnemery commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,284 @@
+                self.hass.config_entries.async_update_entry(
+                    config_entry, data=self.data
+                )
+                await self.hass.config_entries.async_reload(self.context["entry_id"])
```
Why do we call reload here, the config entry has an update listener?

pwheel replied:
Ah. Honestly I was 'inspired' by reviewing the reconfigure flow for other integrations. Thanks

pwheel replied:
sorted

emontnemery commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,284 @@
+        vol.Optional(CONF_API_APP_KEY, default=""): cv.string,
+    }
+)
+STEP_STOP_POINT_DATA_SCHEMA = vol.Schema(
```
Use plural if this schema allows the input of multiple stop points
```suggestion
STEP_STOP_POINTS_DATA_SCHEMA = vol.Schema(
```

emontnemery commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,284 @@
+        config = self.config_entry.options
+        all_stops = deepcopy(config[CONF_STOP_POINTS])
+
+        options_schema = vol.Schema(
+            {
+                vol.Required(CONF_STOP_POINTS, default=all_stops): TextSelector(
+                    TextSelectorConfig(multiple=True)
+                ),
+            }
+        )
```
As above, try a constant schema + add_suggested_values_to_schema

emontnemery commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,284 @@
+                )
+
+        config = self.config_entry.options
+        all_stops = deepcopy(config[CONF_STOP_POINTS])
```
there's no need to deepcopy since we're not mutating?

pwheel replied:
Good catch. Missed to tidy this up when I refactored to use `TextSelector`

pwheel replied:
resolved

emontnemery reviewed (changes requested):
Some more comments. I've reviewed up to config_flow.py

pwheel committed:
Merge branch 'dev' into tfl-integration

pwheel committed:
- Consolidated common error handling
- Moved schemas to constants
- Cleaned up some entries in strings.json

pwheel committed:
- Removed unnecessary reload
- Handled scenario to allow no app key to be specified

pwheel committed:
Migrated off using hass.data to the runtime_data pattern

pwheel committed:
Merge branch 'dev' into tfl-integration

emontnemery reviewed (commented):
Some more comments. I've reviewed up to const.py + a couple of general comments on sensor.py

emontnemery commented on homeassistant/components/tfl/sensor.py:
```diff
@@ -0,0 +1,160 @@
+        self, stop_point_api: stopPoint, name: str, stop_point_id: str, entry_id: str
+    ) -> None:
+        """Initialize the TfL StopPoint sensor."""
+        # super().__init__(coordinator)
```
How come the integration doesn't use a coordinator?

pwheel replied:
TfL doesn't have an API that allows you to retrieve the arrivals for multiple stop points at the same time.

[This is the api used](https://zackaryh8.github.io/tfl-api-wrapper-py/#tflwrapper.stopPoint.getStationArrivals), from the `tflwrapper` python library.

My understanding of the use case for the `DataUpdateCoordinator` ([docs](https://developers.home-assistant.io/docs/integration_fetching_data/)) is when you can have a single API poll for all the entities. That's not the case here. There is a 1:1 mapping between an API call and an entity so it fits under the _Separate polling for each individual entity_ use case.

emontnemery replied:
OK, I guess that makes sense. I'd suggest to add a (very brief) explanation to the sensor class' docstring explaining this.

pwheel replied:
Done in f9d00ea263b8fcc94079b83ebe4622c0fb244d90

pwheel commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,256 @@
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_API_APP_KEY, default=""): cv.string,
```
Good question; will remove

pwheel replied:
Fixed in https://github.com/home-assistant/core/pull/104529/commits/6e429956e6f2ac09ff482c1b39f560b9fe3ef83e

pwheel committed:
upate homeassistant/components/tfl/common.py

Remove empty line

Co-authored-by: Erik Montnemery <erik@montnemery.com>

pwheel committed:
Merge branch 'dev' into tfl-integration

pwheel committed:
- Enabled strict typing for tfl component
- Moved some consts in to sensor.py

pwheel commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,256 @@
+        raise
+
+
+async def validate_stop_points(hass: HomeAssistant, app_key: str, stop_points):
```
Strict type checking has been enabled in https://github.com/home-assistant/core/pull/104529/commits/6e429956e6f2ac09ff482c1b39f560b9fe3ef83e

pwheel commented on homeassistant/components/tfl/const.py:
```diff
@@ -0,0 +1,12 @@
+
+RAW_ARRIVAL_LINE_NAME = "lineName"
+RAW_ARRIVAL_DESTINATION_NAME = "destinationName"
+RAW_ARRIVAL_TIME_TO_STATION = "timeToStation"
```
Done in https://github.com/home-assistant/core/pull/104529/commits/6e429956e6f2ac09ff482c1b39f560b9fe3ef83e

pwheel commented on homeassistant/components/tfl/sensor.py:
```diff
@@ -0,0 +1,160 @@
+                # Due to 255 character limit, the value of the sensor is the next arrival and
+                # the next 3 and full list are provided as attributes
+                self._attr_native_value = arrival_next
+                attributes[ATTR_NEXT_THREE_ARRIVALS] = arrivals_next_3
+                attributes[ATTR_NEXT_ARRIVALS] = arrivals
```
Mostly, you'd be interested in showing the next 3 arrivals at a stop rather than everything, which could be say 12 arrivals.
So as a convenience, it seemed nice to provide an attribute explicitly for the next 3 arrivals so that usages in markdown for example on dashboards don't have to only take the top 3 items and can instead just iterate over the list knowing that it won't be enormous.
Should make life simpler to use it

pwheel replied:
Recommendation from @joostlek to use a coordinator with multiple sensors for each arrival as per `swiss_public_transport` which would result in how this looks changing

emontnemery commented on homeassistant/components/tfl/sensor.py:
```diff
@@ -0,0 +1,162 @@
+
+                # Due to 255 character limit, the value of the sensor is the next arrival and
+                # the next 3 and full list are provided as attributes
+                self._attr_native_value = str(arrival_next)
```
This is not OK.
Make the state something meaningful, most similar integrations makes the state the number of minutes until the next departure.

pwheel replied:
Got it, thank you. Resolved in 9df8a54a34395f43815905a6ebabdfd286df9964

emontnemery commented on homeassistant/components/tfl/sensor.py:
```diff
@@ -0,0 +1,162 @@
+        self._stop_point_api = stop_point_api
+        self._stop_point_id = stop_point_id
+
+    async def async_update(self) -> None:
```
Out of curiosity, why does the sensor focus on arrivals, isn't departure time typically more useful?

pwheel replied:
Interesting question! The API ([example for Oxford Circus](https://api.tfl.gov.uk/StopPoint/940GZZLUOXC/arrivals)) is only providing the `time_to_station` and the API is `/arrivals`.
Which kinda makes sense… in a metro area like London, the tube and the bus will go as soon as it's ready. So the departure time is a function of how many people are getting on and off the tube/bus and can't be predicted. The arrival time is more predictable, and as a traveller you want to be there for the arrival time.

pwheel committed:
Changed validation calls to use a ValidationResult
instead of callings need to catch and handle.
Fixed broken test 'test_async_update_success'

pwheel commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,256 @@
+    _LOGGER.debug("Validating app_key, got categories=%s", categories)
+
+
+async def validate_stop_point(
```
Actioned in 6fbe20eb5fb1a30774444f92d74b804319154995. I ended up using a data class for the response rather than an error dict.

pwheel committed:
Changed the value of the sensor to be the seconds
until the next arrival

pwheel committed:
Added to docstring explaining why a
DataUpdateCoordinator is not used

pwheel committed:
Merge branch 'dev' into tfl-integration

joostlek commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,266 @@
+            description_placeholders=validation_result.description_placeholders,
+        )
+
+    async def async_step_reconfigure(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Reconfigure step allows the app key to be updated."""
+        validation_result = ValidationResult()
+        config_entry = self.hass.config_entries.async_get_entry(
+            self.context["entry_id"]
+        )
+        assert config_entry is not None
+        if user_input is not None:
+            app_key = user_input.get(CONF_API_APP_KEY, "")
+            validation_result = await validate_app_key(self.hass, app_key)
+
+            if not validation_result.errors:
+                # Input is valid, set data
+                self.data[CONF_API_APP_KEY] = app_key
+                self.hass.config_entries.async_update_entry(
+                    config_entry, data=self.data
+                )
+                return self.async_abort(reason="reconfigure_successful")
+
+        return self.async_show_form(
+            step_id="reconfigure",
+            data_schema=self.add_suggested_values_to_schema(
+                RECONFIGURE_DATA_SCHEMA,
+                config_entry.data if user_input is None else user_input,
+            ),
+            errors=validation_result.errors,
+        )
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(
+        config_entry: config_entries.ConfigEntry,
+    ) -> config_entries.OptionsFlow:
```
Let's leave options and reconfigure flow out for now for the scope and size of the PR

pwheel replied:
Sure thing. Keen to get this PR over the line, so happy to do this; makes sense.

pwheel replied:
Done in e591fe286e29db2843826580bd62c19279722202

joostlek commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,266 @@
+    """Handle a config flow for Transport for London."""
+
+    VERSION = 1
+    data: dict[str, Any] = {}
+    options: dict[str, Any] = {}
+
```
Don't create mutable objects like dicts or lists on class level, this means they will be shared between all instances of that class. Rather create them in the constructor so they are created on instance level

pwheel replied:
Doh, of course. Thanks. I'm a Java / Kotlin dev by trade so whilst this is obvious I will blame it on my novice python skills even though the principles are the same ;-)

pwheel replied:
Fixed in 60bce47a177403855332c56c0b5958470540defe

pwheel commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,256 @@
+    return errors, description_placeholders
+
+
+class TfLConfigFlow(ConfigFlow, domain=DOMAIN):
```
@joostlek so moving the configuration of stops out to the options flow?

I'm happy to follow best practices recommendations, but the integration isn't useful without adding some stops. If that matches the user expectation then that's fine. My worry would be is it confusing and more clicks / actions needed to get something running that's useful.

pwheel committed:
Fixing data and options dicts to be instance level
and not class level

pwheel commented on homeassistant/components/tfl/__init__.py:
```diff
@@ -0,0 +1,70 @@
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+type TfLConfigEntry = ConfigEntry[TfLData]
+
+
+@dataclass
+class TfLData:
+    """Runtime data for TfL integration."""
+
+    stop_point_api: stopPoint
```
This was a hedge against the future to make it easier to add stuff in future, but I can take it out

joostlek replied:
We rather make the code for the situation now and expand when needed

pwheel replied:
Done in 7541c53eb1

joostlek commented:
Can you elaborate how the stop system is working?

pwheel committed:
Removing empty config from manifest.json

pwheel commented on homeassistant/components/tfl/manifest.json:
```diff
@@ -0,0 +1,14 @@
+  "iot_class": "cloud_polling",
+  "requirements": ["tflwrapper==1.1.8"],
+  "ssdp": [],
+  "zeroconf": []
```
Done in c9ec61c6ad

pwheel committed:
Removed unnecessary data class for storing runtime data

pwheel committed:
Set entity has name to True

pwheel commented on homeassistant/components/tfl/sensor.py:
```diff
@@ -0,0 +1,160 @@
+        )
+
+
+class StopPointSensor(SensorEntity):
```
Done in 47de6cebd1

pwheel committed:
Async_setup_entry tests now validate that sensors are created.

Sensors were not being created due to missing mocking.
As part of this, updated sensor to use unique_id not entry_id for unique_id

pwheel commented on homeassistant/components/tfl/sensor.py:
```diff
@@ -0,0 +1,175 @@
+
+    try:
+        stop_point_infos = await call_tfl_api(
+            hass, stop_point_api.getByIDs, stop_point_ids, False
```
This code has bugs. The `getByIDs` is not good for this use case.

There are the following problems:
1. Some stop point ids are edge cases such that TfL's stop point hierarchy might return a parent stop point id where the requested stop point id is a child.
2. The API can return the stop points in a different order than requested

So you could want to monitor 3 stop points, but the API might return only 2 because it considers one stop point a child… so this code could get complex to correctly process the response.
Given that all this is trying to do is to obtain the `commonName` for each stop point, it is going to be way simpler to instead use the `getStationArrivals` which takes all that complexity away. The tradeoff is needing to make multiple calls but the simplicity is worth it.

pwheel replied:
Fixed in d8120d885603cc583a491248c5763cca7af6d4a5

pwheel commented:
> Can you elaborate how the stop system is working?

Sure, can try. I'm not sure if you mean conceptually (e.g. what are TfL Stop Points) or technically for this integration, so I'll try to cover both.
Transport for London (TfL) runs the public transit system in London comprosing of: tube/underground, bus, overground (rail), tram, river services.
Anywhere that those services stop and where passengers can board is a _stop point_.
TfL provides an API that exposes the arrivals data for these stop points.

Apps like _Citymapper_ (which also covers the Netherlands) would use this data to display the upcoming arrivals for a given location i.e. stop point.
You can see how that looks on the TfL website and in their API:

1. [Web page for a stop in Camden](https://tfl.gov.uk/bus/stop/490000036R/camden-town-station-camden-street?)
2. That URL has the stop point id in it - 490000036R - which you can use in the `/arrivals` API like [this](https://api.tfl.gov.uk/StopPoint/490000036R/arrivals)

This integration makes that arrival data available in Home Assistant, for the stop(s) you configure.

The integration is just polling the `/arrivals` API (`getStationArrivals` in the `tflwrapper` library) and exposing the data.

Hope that makes sense; can also discuss on discourse.

pwheel committed:
Handling more complicating stop point info responses.

Fixed the scenario where the stop point info call
can return hierarchical stop point data.

pwheel committed:
Simplify the PR by removing options and reconfigure flow for now

pwheel commented:
Hi  @emontnemery , @joostlek  - appreciate your feedback so far. Can you take another look please? There's no feedback still needing action from my side. Thanks

pwheel committed:
Fixing pylint error from CI/CD

pwheel commented on homeassistant/components/tfl/common.py:
```diff
@@ -0,0 +1,34 @@
+from homeassistant.exceptions import HomeAssistantError
+
+
+async def call_tfl_api[*_Ts, _T](
+    hass: HomeAssistant, target: Callable[[*_Ts], _T], *args: *_Ts
+) -> _T:
+    """Execute a call to TfL using the tflwrapper library using common error handling."""
+    try:
+        return_val = await hass.async_add_executor_job(target, *args)
+    except HTTPError as exception:
+        # TfL's API returns a 429 if you pass an invalid app_key, but we also check
+        # for other reasonable error codes in case their behaviour changes
+        error_code = exception.code
+        if error_code in (429, 401, 403):
+            raise InvalidAuth from exception
+        raise
+    except URLError as exception:
+        raise CannotConnect from exception
+
+    return return_val
+
+
+class CannotConnect(HomeAssistantError):
+    """Error to indicate we cannot connect."""
+
+
+class InvalidAuth(HomeAssistantError):
+    """Error to indicate there is invalid auth."""
```
The library is providing just a thin wrapper around the TfL REST API, so doesn't make an attempt to interpret any error codes - it just lets them bubble up. From the library perspective, that's a design choice.

From the integration perspective, it's reasonable to hedge our bets slightly as the choice of which 4xx error code is returned by the TfL REST API for invalid auth isn't great (429) - but perhaps in future that changes to 401 and if so we don't want to have to do a bug fix and update the integration.
And that then drives centralising the logic here rather than duplicating it at all the call sites.

The library wants to keep it simple and not assume that 429 is invalid auth and have to release a bug fix if the TfL REST API starts using 429 properly.
I think the library is making reasonable choices.

But yeah - ideally this code wouldn't need to exist.

pwheel commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,180 @@
+STEP_STOP_POINTS_DATA_SCHEMA = vol.Schema(
+    {vol.Required(CONF_STOP_POINTS): TextSelector(TextSelectorConfig(multiple=True))}
+)
+RECONFIGURE_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_API_APP_KEY): cv.string,
+    }
+)
+OPTIONS_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_STOP_POINTS): TextSelector(TextSelectorConfig(multiple=True)),
+    }
+)
```
Ah, nice. This was missed when I took out the reconfigure flow to keep the PR small. Thanks.

pwheel replied:
Fixed in a812ebe151c2156725547a0913bb2040f7b6c462

pwheel committed:
Removed unused code after removing the reconfigure flow

pwheel commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,180 @@
+        vol.Optional(CONF_API_APP_KEY): cv.string,
+    }
+)
+STEP_STOP_POINTS_DATA_SCHEMA = vol.Schema(
+    {vol.Required(CONF_STOP_POINTS): TextSelector(TextSelectorConfig(multiple=True))}
+)
```
There's a lot :-)

This could be improved over time perhaps by using the search functionality to do something like autosuggest. That API is like this - https://api.tfl.gov.uk/StopPoint/Search/wood%20green
But then you have stops with the same name but going in different directions so how to display that is maybe tricky.

The intent here was to keep it simple for this first cut - just allow users to manually enter the stop point ids to keep the PR small.

pwheel committed:
Removed unneeded update listener left over from reconfigure flow

pwheel commented on homeassistant/components/tfl/__init__.py:
```diff
@@ -0,0 +1,62 @@
+    """Handle options update."""
+    _LOGGER.debug("update_listener called")
+    await hass.config_entries.async_reload(entry.entry_id)
+
```
Fixed in 0ddad2f850

pwheel committed:
Tidied up the try block for validating the app key

pwheel commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,180 @@
+        stop_point_api = stopPoint(app_key)
+        # Make a random, cheap, call to the API to validate the app_key
+        categories = await call_tfl_api(hass, stop_point_api.getCategories)
+        _LOGGER.debug("Validating app_key, got categories=%s", categories)
```
Resolved in 16ceee6ae9

pwheel committed:
Tidied up the try block for validating a stop point.

pwheel commented on homeassistant/components/tfl/config_flow.py:
```diff
@@ -0,0 +1,180 @@
+        arrivals = await call_tfl_api(
+            hass, stop_point_api.getStationArrivals, stop_point
+        )
+        _LOGGER.debug("Got for stop_point=%s, arrivals=%s", stop_point, arrivals)
```
Resolved in 86266ada81

pwheel committed:
Camel case to snake case

pwheel commented on homeassistant/components/tfl/sensor.py:
```diff
@@ -0,0 +1,184 @@
+    async_add_entities(devices, True)
+
+
+def __getCommonName(stop_point_id: str, stop_point_info: dict[str, typing.Any]) -> str:
```
Fixed in 0127e9f852

pwheel committed:
Remove commented out import

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

pwheel committed:
Organised imports

pwheel commented on homeassistant/components/tfl/sensor.py:
```diff
@@ -0,0 +1,184 @@
+        self._stop_point_api = stop_point_api
+        self._stop_point_id = stop_point_id
+
+    async def async_update(self) -> None:
```
Ah, I see how that works. Yes, that is quite a nice pattern. Thanks for referencing the example.
I can make that refactor, a little bigger than I was hoping for but that's alright

pwheel replied:
Hmm, actually it's more complex. The `swiss_public_transport` integration is doing a point A to point B connection, so there is no additional information associated with each connection.

That's not the case for a TfL stop point. There may be say 3 different buses going through a stop point which all go in the same direction up to a point but have different final destinations. Similarly, for a tube / metro stop point there may be 2 different lines running through a stop point (a tube station platform) going in the same direction up to a point; or the same tube line but with different destinations (either not going to the end of the line or some tube lines have multiple destinations).

@joostlek how would you recommend encoding that information? I could create multiple sensors for each arrival, one sensor each for: time to station, line name, destination, arrival time… but that seems unwieldy.
It would end up with the following sensors:
1. Arrival, Arrival +1, Arrival +2 (timestamp)
2. Time To Station, Time To Station +1, Time To Station +2 (duration)
3. Line Name, Line Name +1, Line Name +2
4. Destination, Destination +1, Destination +2 

Would it make sense to have this extra information as attributes on an arrival sensor?
So you might have:
1. Arrival, Arrival +1, Arrival +2 - each of these has the following attributes: destination, line name, time to station

pwheel replied:
Adding extra state attributes makes more sense to me so will go that route
